pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,777 @@
+	ovsExternalIDEntityNamespace = "entity-namespace"
+	ovsExternalIDIP              = "ip"
+	eeSplitter                   = ","
+	reserveRuleSplitter          = ":"
+)
+
+var (
+	keyFunc      = cache.DeletionHandlingMetaNamespaceKeyFunc
+	splitKeyFunc = cache.SplitMetaNamespaceKey
+)
+
+type ExternalNodeController struct {
+	ovsBridgeClient          ovsconfig.OVSBridgeClient
+	ovsctlClient             ovsctl.OVSCtlClient
+	ofClient                 openflow.Client
+	externalNodeInformer     cache.SharedIndexInformer
+	externalNodeLister       enlister.ExternalNodeNamespaceLister
+	externalNodeListerSynced cache.InformerSynced
+	queue                    workqueue.RateLimitingInterface
+	ifaceStore               interfacestore.InterfaceStore
+	syncedExternalNodes      cache.Store
+	// nodeUpdateNotifier is used for notifying updates of local ExternalNode to NetworkPolicyController.
+	nodeUpdateNotifier channel.Notifier
+	nodeName           string
+	namespace          string
+	reservedHostPorts  []reserveHostPort
+	policyBypassRules  []string
+}
+
+type reserveHostPort struct {
+	// The protocol (TCP, UDP, or SCTP) which traffic must match.
+	protocol binding.Protocol
+	// The dst port on the given protocol.
+	port uint16
+	// The remote IP to which is reserved.
+	ip net.IP
+	// The remote CIDR to which is reserved.
+	ipnet   *net.IPNet
+	ingress bool
+}
+
+func NewExternalNodeController(ovsBridgeClient ovsconfig.OVSBridgeClient, ofClient openflow.Client, externalNodeInformer eninformer.ExternalNodeInformer,
+	ifaceStore interfacestore.InterfaceStore, nodeUpdateNotifier channel.Notifier, namespace string, policyBypassRules []string) (*ExternalNodeController, error) {
+	c := &ExternalNodeController{
+		ovsBridgeClient:          ovsBridgeClient,
+		ovsctlClient:             ovsctl.NewClient(ovsBridgeClient.GetBridgeName()),
+		ofClient:                 ofClient,
+		externalNodeInformer:     externalNodeInformer.Informer(),
+		externalNodeLister:       externalNodeInformer.Lister().ExternalNodes(namespace),
+		externalNodeListerSynced: externalNodeInformer.Informer().HasSynced,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "externalNode"),
+		ifaceStore:               ifaceStore,
+		syncedExternalNodes:      cache.NewStore(keyFunc),
+		nodeUpdateNotifier:       nodeUpdateNotifier,
+		policyBypassRules:        policyBypassRules,
+	}
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		return nil, err
+	}
+	c.nodeName = nodeName
+	c.namespace = namespace
+	c.externalNodeInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueueExternalNodeAdd,
+			UpdateFunc: c.enqueueExternalNodeUpdate,
+			DeleteFunc: c.enqueueExternalNodeDelete,
+		},
+		resyncPeriod)
+
+	return c, nil
+}
+
+// Run will create defaultWorkers workers (goroutines) which will process the ExternalNode events from the work queue.
```

Comment:
It runs only a single worker?

Reply:
I will update the comment here. Since agent will only process its ExternalNode, we think one worker is enough here.